lib/Conversion/AffineToPipeline/AffineToPipeline.cpp

```diff
@@ -264,14 +270,46 @@ LogicalResult AffineToPipeline::populateOperatorTypes(
           problem.setLinkedOperatorType(combOp, combOpr);
           return WalkResult::advance();
         })
-        .Case<AffineLoadOp, AffineStoreOp, memref::LoadOp, memref::StoreOp>(
-            [&](Operation *seqOp) {
-              // Some known sequential ops. In certain cases, reads may be
-              // combinational in Calyx, but taking advantage of that is left as
-              // a future enhancement.
-              problem.setLinkedOperatorType(seqOp, seqOpr);
-              return WalkResult::advance();
-            })
+        .Case<AddIOp, CmpIOp>([&](Operation *seqOp) {
+          // These ops need to be sequential for now because we do not
+          // have enough information to chain them together yet.
```

Comment:
They are combinational, but if we have a long chain of add ops in sequence without being broken by a sequential op we will create a very long combinational chain that will severely hurt the max frequency. What we need is the delay information provided by the ChainingProblem to determine how many ops we can chain together before we will hurt the target clock frequency. We don't have this information for now, so the safest thing to do is just treat these ops as sequential when scheduling. Even if we had something like a ModuloChainingProblem it would not be useful at all without operator delay characterization.

The ops left as 0 cycle are ones that have very very low or zero effect on the frequency, such as index cast operations.

Reply:
I see, then I think this makes sense to get something working with reasonable FMAX. BTW, are you targeting FPGA or ASIC (or both?). Longer term, I'd love for the decisions made in this code to be parameterized by a proper "target operator library" that includes information like combinational delays. Getting something that's a reasonable default for Calyx and some target device in the short term has always been the goal of this pass :) And we can improve/generalize from there.

Reply:
Eventually we want to target both, but mostly FPGA for now. Yes, that decision falls out naturally from having a target operator library with operator delays (and a target frequency to try to meet). 

There is a ton of generalization that comes naturally from having a target operator library, but we would need a good separate program to characterize the operator delays for a specific device. It's actually even a bit more complicated than that because we need to characterize each operator for each possible bit-width (1-128 or 256) and build a function from bit-width to operator delay (otherwise the operator library will become huge).

Reply:
> if we had something like a ModuloChainingProblem

Let me know when you need that, it's easy to define, and implement a corresponding scheduler (chaining support composes well with LP-based schedulers).